File: frontend/src/features/checkmarx/services/checkmarxScan.ts
import { apiClient } from '@/shared/api/serverClient';

export interface CheckmarxScanJob {
  id: string;
  status: string;
  current_step?: string | null;
  progress: number;
  project_name: string;
  source_filename: string;
  checkmarx_base_url?: string | null;
  checkmarx_project_id?: string | null;
  scan_id?: string | null;
  totals_json?: string | null;
  error_message?: string | null;
  created_at?: string;
  started_at?: string | null;
  completed_at?: string | null;
  results_count: number;
}

export interface CheckmarxScanResult {
  id: string;
  scan_id: string;
  path_id: string;
  vulnerability: string;
  type: string;
  severity?: number | null;
  url: string;
  ai_judgement?: boolean | null;
  ai_reason?: string | null;
}

export async function createCheckmarxScan(params: {
  file: File;
  projectName: string;
  username: string;
  password: string;
  baseUrl?: string;
}): Promise<CheckmarxScanJob> {
  const formData = new FormData();
  formData.append('file', params.file);
  formData.append('project_name', params.projectName);
  formData.append('username', params.username);
  formData.append('password', params.password);
  if (params.baseUrl?.trim()) {
    formData.append('base_url', params.baseUrl.trim());
  }

  const response = await apiClient.post<CheckmarxScanJob>('/checkmarx/scans', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  });
  return response.data;
}

export async function getCheckmarxScan(jobId: string): Promise<CheckmarxScanJob> {
  const response = await apiClient.get<CheckmarxScanJob>(`/checkmarx/scans/${jobId}`);
  return response.data;
}

export async function listCheckmarxScans(): Promise<CheckmarxScanJob[]> {
  const response = await apiClient.get<CheckmarxScanJob[]>('/checkmarx/scans');
  return response.data;
}

export async function listCheckmarxScanResults(jobId: string): Promise<CheckmarxScanResult[]> {
  const response = await apiClient.get<CheckmarxScanResult[]>(`/checkmarx/scans/${jobId}/results`);
  return response.data;
}

export async function downloadCheckmarxExcel(jobId: string): Promise<Blob> {
  const response = await apiClient.get(`/checkmarx/scans/${jobId}/export.xlsx`, {
    responseType: 'blob',
  });
  return response.data;
}

File: backend/app/models/checkmarx_scan.py
import uuid

from sqlalchemy import Boolean, Column, DateTime, ForeignKey, Integer, String, Text
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.db.base import Base


class CheckmarxScanJob(Base):
    __tablename__ = "checkmarx_scan_jobs"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    created_by = Column(String, ForeignKey("users.id"), nullable=False, index=True)

    status = Column(String, default="pending", nullable=False, index=True)
    current_step = Column(String, nullable=True)
    progress = Column(Integer, default=0, nullable=False)

    project_name = Column(String, nullable=False)
    source_filename = Column(String, nullable=False)
    checkmarx_base_url = Column(String, nullable=True)
    checkmarx_project_id = Column(String, nullable=True)
    scan_id = Column(String, nullable=True, index=True)

    totals_json = Column(Text, nullable=True)
    error_message = Column(Text, nullable=True)

    started_at = Column(DateTime(timezone=True), nullable=True)
    completed_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())

    creator = relationship("User", foreign_keys=[created_by])
    results = relationship("CheckmarxScanResult", back_populates="job", cascade="all, delete-orphan")


class CheckmarxScanResult(Base):
    __tablename__ = "checkmarx_scan_results"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    job_id = Column(String, ForeignKey("checkmarx_scan_jobs.id", ondelete="CASCADE"), nullable=False, index=True)

    scan_id = Column(String, nullable=False, index=True)
    path_id = Column(String, nullable=False, index=True)
    vulnerability = Column(String, nullable=False)
    type = Column(String, nullable=False)
    severity = Column(Integer, nullable=True)
    url = Column(Text, nullable=False)

    ai_judgement = Column(Boolean, nullable=True)
    ai_reason = Column(Text, nullable=True)

    raw_result = Column(Text, nullable=True)
    workflow_response = Column(Text, nullable=True)

    created_at = Column(DateTime(timezone=True), server_default=func.now())

    job = relationship("CheckmarxScanJob", back_populates="results")
